File: PB004.fs/Algorithm.fs
namespace ProjectEuler

type public Algorithm() = class
    let isPalindrome (str:string) = 
        let n = str.Length
        match str with 
        | "" -> true
        | _ -> seq {0..(n >>> 1)}
            |> Seq.exists (fun i -> str.[i] <> str.[n - i - 1])
            |> not

    let run =
        seq {for i in 100..999 do
                for j in i..999 -> i * j}
        |> Seq.sortDescending
        |> Seq.map string
        |> Seq.filter isPalindrome
        |> Seq.head

    interface IAlgorithm with
        member this.Compute(): string = run |> string
        member this.Prepare(): bool = true
    end
end

File: PB004.cs/Algorithm.cs
namespace ProjectEuler
{
    public class Algorithm : IAlgorithm
    {
        public string Compute()
        {
            int max = int.MinValue;
            for (int i = 100; i <= 999; i++)
                for (int j = i; j <= 999; j++)
                    if (TroyMath.isPalindrome((i * j).ToString()))
                        if (i * j > max)
                            max = i * j;
            return max.ToString();
        }

        public bool Prepare()
        {
            return true;
        }
    }
}
